[CI] EsqlDisruptionIT testExtractFields failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/ing5bk2dhs36u/tests/:x-pack:plugin:esql:internalClusterTest/org.elasticsearch.xpack.esql.action.EsqlDisruptionIT/testExtractFields


**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:internalClusterTest' --tests "org.elasticsearch.xpack.esql.action.EsqlDisruptionIT.testExtractFields" -Dtests.seed=70F102B72FD3E64A -Dtests.locale=ga-IE -Dtests.timezone=America/Belem -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.esql.action.EsqlDisruptionIT#testExtractFields`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testExtractFields'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.action.EsqlDisruptionIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: <[Doc[val=0, tag=tag-5], Doc[val=1, tag=tag-1], Doc[val=2, tag=tag-93], Doc[val=3, tag=tag-75], Doc[val=4, tag=tag-12], Doc[val=5, tag=tag-37], Doc[val=6, tag=tag-77], Doc[val=7, tag=tag-14], Doc[val=8, tag=tag-43], Doc[val=9, tag=tag-46]]>
     but: was <[Doc[val=0, tag=tag-5], Doc[val=1, tag=tag-93], Doc[val=2, tag=tag-93], Doc[val=3, tag=tag-75], Doc[val=4, tag=tag-77], Doc[val=5, tag=tag-77], Doc[val=6, tag=tag-14], Doc[val=7, tag=tag-43], Doc[val=8, tag=tag-43], Doc[val=9, tag=tag-46]]>

  at __randomizedtesting.SeedInfo.seed([70F102B72FD3E64A:65383D3174BB6B79]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.xpack.esql.action.EsqlActionIT.testExtractFields(EsqlActionIT.java:840)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

--- a/server/src/main/java/org/elasticsearch/index/mapper/BlockDocValuesReader.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/BlockDocValuesReader.java
@@ -558,7 +558,8 @@ public abstract class BlockDocValuesReader implements BlockLoader.AllReader {
         private BlockLoader.Block readSingleDoc(BlockFactory factory, int docId) throws IOException {
             if (ordinals.advanceExact(docId)) {
                 BytesRef v = ordinals.lookupOrd(ordinals.ordValue());
-                return factory.constantBytes(v);
+                // the returned BytesRef can be reused
+                return factory.constantBytes(BytesRef.deepCopyOf(v));
             } else {
                 return factory.constantNulls();
             }

--- a/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperatorTests.java
+++ b/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperatorTests.java
@@ -78,6 +78,7 @@ import org.junit.After;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -111,6 +112,7 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
 
     private Directory directory = newDirectory();
     private IndexReader reader;
+    private static final Map<Integer, String> keyToTags = new HashMap<>();
 
     @After
     public void closeIndex() throws IOException {
@@ -146,10 +148,10 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
     protected SourceOperator simpleInput(BlockFactory blockFactory, int size) {
         // The test wants more than one segment. We shoot for 10.
         int commitEvery = Math.max(1, (int) Math.ceil((double) size / 10));
-        return simpleInput(driverContext(), size, commitEvery);
+        return simpleInput(driverContext(), size, commitEvery, randomPageSize());
     }
 
-    private SourceOperator simpleInput(DriverContext context, int size, int commitEvery) {
+    private SourceOperator simpleInput(DriverContext context, int size, int commitEvery, int pageSize) {
         try {
             initIndex(size, commitEvery);
         } catch (IOException e) {
@@ -160,13 +162,14 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
             ctx -> new MatchAllDocsQuery(),
             DataPartitioning.SHARD,
             randomIntBetween(1, 10),
-            size,
+            pageSize,
             LuceneOperator.NO_LIMIT
         );
         return luceneFactory.get(context);
     }
 
     private void initIndex(int size, int commitEvery) throws IOException {
+        keyToTags.clear();
         try (
             IndexWriter writer = new IndexWriter(
                 directory,
@@ -181,9 +184,8 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
                 doc.add(new SortedNumericDocValuesField("short", (short) d));
                 doc.add(new SortedNumericDocValuesField("byte", (byte) d));
                 doc.add(new SortedNumericDocValuesField("long", d));
-                doc.add(
-                    new KeywordFieldMapper.KeywordField("kwd", new BytesRef(Integer.toString(d)), KeywordFieldMapper.Defaults.FIELD_TYPE)
-                );
+                String tag = keyToTags.computeIfAbsent(d, k -> "tag-" + randomIntBetween(1, 5));
+                doc.add(new KeywordFieldMapper.KeywordField("kwd", new BytesRef(tag), KeywordFieldMapper.Defaults.FIELD_TYPE));
                 doc.add(new StoredField("stored_kwd", new BytesRef(Integer.toString(d))));
                 doc.add(new StoredField("stored_text", Integer.toString(d)));
                 doc.add(new SortedNumericDocValuesField("bool", d % 2 == 0 ? 1 : 0));
@@ -293,6 +295,37 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
             ),
             Block.MvOrdering.UNORDERED
         );
+    }
+
+    public void testManySingleDocPages() {
+        DriverContext driverContext = driverContext();
+        int numDocs = between(10, 100);
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(driverContext, numDocs, numDocs, 1));
+        Randomness.shuffle(input);
+        List<Operator> operators = new ArrayList<>();
+        Checks checks = new Checks(Block.MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING);
+        FieldCase testCase = new FieldCase(
+            new KeywordFieldMapper.KeywordFieldType("kwd"),
+            checks::tags,
+            StatusChecks::keywordsFromDocValues
+        );
+        operators.add(
+            new ValuesSourceReaderOperator.Factory(
+                List.of(testCase.info, fieldInfo(docValuesNumberField("key", NumberFieldMapper.NumberType.INTEGER))),
+                List.of(new ValuesSourceReaderOperator.ShardContext(reader, () -> SourceLoader.FROM_STORED_SOURCE)),
+                0
+            ).get(driverContext)
+        );
+        List<Page> results = drive(operators, input.iterator(), driverContext);
+        assertThat(results, hasSize(input.size()));
+        for (Page page : results) {
+            assertThat(page.getBlockCount(), equalTo(3));
+            IntVector keys = page.<IntBlock>getBlock(2).asVector();
+            for (int p = 0; p < page.getPositionCount(); p++) {
+                int key = keys.getInt(p);
+                testCase.checkResults.check(page.getBlock(1), p, key);
+            }
+        }
     }
 
     public void testEmpty() {
@@ -585,7 +618,7 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
         r.add(new FieldCase(new BooleanFieldMapper.BooleanFieldType("bool"), checks::bools, StatusChecks::boolFromDocValues));
         r.add(new FieldCase(new BooleanFieldMapper.BooleanFieldType("mv_bool"), checks::mvBools, StatusChecks::mvBoolFromDocValues));
         r.add(new FieldCase(new BooleanFieldMapper.BooleanFieldType("missing_bool"), checks::constantNulls, StatusChecks::constantNulls));
-        r.add(new FieldCase(new KeywordFieldMapper.KeywordFieldType("kwd"), checks::strings, StatusChecks::keywordsFromDocValues));
+        r.add(new FieldCase(new KeywordFieldMapper.KeywordFieldType("kwd"), checks::tags, StatusChecks::keywordsFromDocValues));
         r.add(
             new FieldCase(
                 new KeywordFieldMapper.KeywordFieldType("mv_kwd"),
@@ -611,7 +644,7 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
         r.add(
             new FieldCase(
                 textFieldWithDelegate("text_with_delegate", new KeywordFieldMapper.KeywordFieldType("kwd")),
-                checks::strings,
+                checks::tags,
                 StatusChecks::textWithDelegate
             )
         );
@@ -678,6 +711,11 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
         void strings(Block block, int position, int key) {
             BytesRefVector keywords = ((BytesRefBlock) block).asVector();
             assertThat(keywords.getBytesRef(position, new BytesRef()).utf8ToString(), equalTo(Integer.toString(key)));
+        }
+
+        void tags(Block block, int position, int key) {
+            BytesRefVector keywords = ((BytesRefBlock) block).asVector();
+            assertThat(keywords.getBytesRef(position, new BytesRef()).utf8ToString(), equalTo(keyToTags.get(key)));
         }
 
         void bools(Block block, int position, int key) {
@@ -1285,7 +1323,7 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
 
     private void testSequentialStoredFields(boolean sequential, int docCount) {
         DriverContext driverContext = driverContext();
-        List<Page> source = CannedSourceOperator.collectPages(simpleInput(driverContext, docCount, docCount));
+        List<Page> source = CannedSourceOperator.collectPages(simpleInput(driverContext, docCount, docCount, docCount));
         assertThat(source, hasSize(1)); // We want one page for simpler assertions, and we want them all in one segment
         assertTrue(source.get(0).<DocBlock>getBlock(0).asVector().singleSegmentNonDecreasing());
         Operator op = new ValuesSourceReaderOperator.Factory(

--- a/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperatorTests.java
+++ b/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperatorTests.java
@@ -146,9 +146,11 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
 
     @Override
     protected SourceOperator simpleInput(BlockFactory blockFactory, int size) {
-        // The test wants more than one segment. We shoot for 10.
-        int commitEvery = Math.max(1, (int) Math.ceil((double) size / 10));
-        return simpleInput(driverContext(), size, commitEvery, randomPageSize());
+        return simpleInput(driverContext(), size, commitEvery(size), randomPageSize());
+    }
+
+    private int commitEvery(int numDocs) {
+        return Math.max(1, (int) Math.ceil((double) numDocs / 10));
     }
 
     private SourceOperator simpleInput(DriverContext context, int size, int commitEvery, int pageSize) {
@@ -300,7 +302,7 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
     public void testManySingleDocPages() {
         DriverContext driverContext = driverContext();
         int numDocs = between(10, 100);
-        List<Page> input = CannedSourceOperator.collectPages(simpleInput(driverContext, numDocs, numDocs, 1));
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(driverContext, numDocs, between(1, numDocs), 1));
         Randomness.shuffle(input);
         List<Operator> operators = new ArrayList<>();
         Checks checks = new Checks(Block.MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING);
@@ -473,7 +475,8 @@ public class ValuesSourceReaderOperatorTests extends OperatorTestCase {
 
     private void testLoadAllStatus(boolean allInOnePage) {
         DriverContext driverContext = driverContext();
-        List<Page> input = CannedSourceOperator.collectPages(simpleInput(driverContext.blockFactory(), between(100, 5000)));
+        int numDocs = between(100, 5000);
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(driverContext, numDocs, commitEvery(numDocs), numDocs));
         assertThat(reader.leaves(), hasSize(10));
         assertThat(input, hasSize(10));
         List<FieldCase> cases = infoAndChecksForEachType(Block.MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING);

